parl/framework/computation_task.py

```diff
@@ -0,0 +1,180 @@
+
+    def _define_program(self):
+        self.learn_program = fluid.Program()
+        self.predict_program = fluid.Program()
+
+        def _get_next_specs(specs):
+            return [("next_" + spec[0], spec[1]) for spec in specs]
+
+        def _select_data(data_layer_dict, specs):
+            return {name: data_layer_dict[name] for name, _ in specs}
+
+        input_specs = self.alg.get_input_specs()
+        state_specs = self.alg.get_state_specs()
+        next_input_specs = _get_next_specs(input_specs)
+        next_state_specs = _get_next_specs(state_specs)
+        action_specs = self.alg.get_action_specs()
+        reward_specs = self.alg.get_reward_specs()
+        episode_end_specs = [("episode_end", dict(shape=[1]))]
+
+        self.action_names = sorted([name for name, _ in action_specs])
+        self.state_names = sorted([name for name, _ in state_specs])
+
+        with fluid.program_guard(self.predict_program):
+            data_layer_dict = self._create_data_layers(input_specs)
+            data_layer_dict.update(self._create_data_layers(state_specs))
+            self.predict_feed_names = sorted(data_layer_dict.keys())
+
+            inputs = _select_data(data_layer_dict, input_specs)
+            states = _select_data(data_layer_dict, state_specs)
+
+            ### call alg predict()
+            pred_actions, pred_states = self.alg.predict(inputs, states)
+            self.predict_fetch = [pred_actions, pred_states]
+
+        with fluid.program_guard(self.learn_program):
+            data_layer_dict = self._create_data_layers(input_specs)
+            data_layer_dict.update(self._create_data_layers(state_specs))
+            data_layer_dict.update(self._create_data_layers(next_input_specs))
+            data_layer_dict.update(self._create_data_layers(next_state_specs))
+            data_layer_dict.update(self._create_data_layers(action_specs))
+            data_layer_dict.update(self._create_data_layers(reward_specs))
+            data_layer_dict.update(self._create_data_layers(episode_end_specs))
+            self.learn_feed_names = sorted(data_layer_dict.keys())
+
+            inputs = _select_data(data_layer_dict, input_specs)
+            states = _select_data(data_layer_dict, state_specs)
+            next_inputs = _select_data(data_layer_dict, next_input_specs)
+            next_states = _select_data(data_layer_dict, next_state_specs)
+            actions = _select_data(data_layer_dict, action_specs)
+            rewards = _select_data(data_layer_dict, reward_specs)
+            episode_end = _select_data(data_layer_dict, episode_end_specs)
+
+            ## call alg learn()
+            ### TODO: implement a recurrent layer to strip the sequence information
+            self.cost = self.alg.learn(inputs, next_inputs, states,
+                                       next_states, episode_end, actions,
+                                       rewards)
+
+    def predict(self, inputs, states=dict()):
```

Comment:
In this function, it seems that there is no need to distinguish inputs and states, we can just use one dict "inputs"

Reply:
No, this will cause inconsistency between the concept of inputs here in this function and elsewhere in the other parts of code. The conceputal consistency is important in my opinion.